### PR TITLE
SCRUM-86 RF-5.2: Implement Command Strategies

### PR DESCRIPTION
**Pull request by TQT279**
# 🚀 Refactor: Implement Strategy Pattern for Chat Commands & Refactor Handler

## 📝 Description
This PR refactors the legacy monolithic command handling logic in `ChatWebSocketHandler` into a scalable **Strategy Pattern**.

Previously, all command logic (Login, Select, List, etc.) was hardcoded inside a large `switch-case` block. This PR decouples specific command logic into separate classes implementing `IChatCommand`, orchestrated by a central `CommandExecutor`.

Additionally, this PR introduces a strict `CommandParserService` using Regex and achieves **100% Unit Test coverage** for the command layer.

### 1. Architecture Refactoring (Strategy Pattern)
- **`ChatWebSocketHandler`**: Removed legacy `switch-case` logic. Now delegates requests to `CommandExecutor`.
- **`CommandExecutor`**: Acts as a dispatcher, routing parsed commands to the appropriate `IChatCommand` implementation.
- **`IChatCommand` Interface**: Standardized interface for all chat commands (`execute()` method).

### 2. New Components
- **`CommandParserService`**: Implemented strict Regex parsing (`^(/[a-zA-Z0-9]+)(?:\\s+(.*))?$`) to separate command type and arguments.
- **Command Implementations**:
    - `LoginCommand`: Handles user registration with validation (alphanumeric, 3-20 chars).
    - `SelectCommand`: Handles private chat switching with strict argument validation (no spaces).
    - `ListCommand`: Filters and lists online users (hiding anonymous ones).
    - `HelpCommand`, `LeaveCommand`, `ExitCommand`, `UnknownCommand`.
- **Utilities**:
    - **`CommandType`**: Converted to Rich Enum with **O(1) static HashMap lookup** (Case-Insensitive).
    - **`SystemResponseSender`**: Centralized response logic (Error/System messages) to adhere to DRY principles.
    - **`MessageConstants`**: Centralized all String literals and **ANSI Color codes** for consistent UI.

### 3. Bug Fixes
- **ExitCommand**: Fixed `NullPointerException` where the server would broadcast "Goodbye, null!" for anonymous users.
- **SelectCommand**: Added validation to prevent selecting usernames containing spaces.

### 4. Testing (Unit Tests)
- Added **JUnit 5 & Mockito** tests covering **100% of business logic**:
    - `CommandParserServiceTest`: Verifies regex logic, edge cases (null/empty), and case-insensitivity.
    - `LoginCommandTest`: Mocks `SessionRegistry` to test duplicate username handling and validation.
    - `SelectCommandTest`, `ExitCommandTest`, `ListCommandTest`, etc.

## 🔗 Related Issues
- Refers to Sprint: **Sprint 5 - Command Pattern Refactoring**

### CODE REVIEW

File: src/main/java/org/example/VChatTalk/command/MessageConstants.java
Comment (Toffu-BlankieChoccy) [inline src/main/java/org/example/VChatTalk/command/MessageConstants.java:1]: Can you document this file to the Jira too? Briefly give a reason the purpose of this file

Comment (TQT279) [reply src/main/java/org/example/VChatTalk/command/MessageConstants.java:1]: Yes, i can. I will add the doc later

Comment (TQT279) [reply src/main/java/org/example/VChatTalk/command/MessageConstants.java:1]: i need permission, so i just send a request to join page jira now need approval to add doc

File: src/test/java/org/example/VChatTalk/command/CommandIntegrationTest.java [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+                new SelectCommand(sessionRegistry, responder),
+                new ListCommand(sessionRegistry, responder),
+                new HelpCommand(responder),
+                new LeaveCommand(sessionRegistry, responder),
+                new ExitCommand(sessionRegistry, responder),
+                new UnknownCommand(responder)
+        );
+
+        // 3. Initialize Executor
+        commandExecutor = new CommandExecutor(commands);
```
[the remark this answers is not in the document]
Comment (TQT279) [reply src/test/java/org/example/VChatTalk/command/CommandIntegrationTest.java:54]: adding testConstructor_DuplicateCommands() pass

File: src/test/java/org/example/VChatTalk/command/implement_command/HelpCommandTest.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+        String sentMessage = messageCaptor.getValue();
+
+        // Check message content
+        // Must have a title
+        assertTrue(sentMessage.contains("Available commands:"));
+
+        // Must contain important commands (Randomly check a few commands)
+        assertTrue(sentMessage.contains("/help"));
+        assertTrue(sentMessage.contains("/join"));
```
[the remark this answers is not in the document]
Comment (TQT279) [reply src/test/java/org/example/VChatTalk/command/implement_command/HelpCommandTest.java]: adding comment sprint 6

File: src/main/java/org/example/VChatTalk/command/service/CommandParserService.java [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+public class CommandParserService {
+
+    // Regex Explanation:
+    // ^                 : Start of string
+    // (/[a-zA-Z0-9]+)   : Group 1 - Command (Must start with '/' followed by alphanumeric characters)
+    // (?:\\s+(.*))?     : Non-capturing Group (Optional part for argument)
+    //    \\s+           : At least one whitespace separator
+    //    (.*)           : Group 2 - Argument (Captures the rest of the string)
+    // $                 : End of string
+    private static final Pattern COMMAND_PATTERN = Pattern.compile("^(/[a-zA-Z0-9]+)(?:\\s+(.*))?$");
```
[the remark this answers is not in the document]
Comment (TQT279) [reply src/main/java/org/example/VChatTalk/command/service/CommandParserService.java:20]: in this project will not having command like /list_all and similar commands, if feature have it will update this line, not now

File: src/main/java/org/example/VChatTalk/util/SystemResponseSender.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    }
+
+    public void sendError(WebSocketSession session, String content) throws IOException {
+        send(session, MessageType.ERROR, content);
+    }
+
+    private void send(WebSocketSession session, MessageType type, String content) throws IOException {
+        if (session == null || !session.isOpen()) {
+            return;
+        }
```
[the remark this answers is not in the document]
Comment (TQT279) [reply src/main/java/org/example/VChatTalk/util/SystemResponseSender.java:36]: adding try catch and
log.warn("Cannot send '{}' message. Session is null or closed. SessionID: {}", type, (session != null ? session.getId() : "null"));

File: src/test/java/org/example/VChatTalk/command/implement_command/SelectCommandTest.java [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+    @Test
+    @DisplayName("Fail if username contains spaces")
+    void testExecute_SpaceInUsername() throws IOException {
+        when(session.getId()).thenReturn("sess-1");
+        when(sessionRegistry.isUserRegistered("sess-1")).thenReturn(true);
+
+        CommandResult result = new CommandResult(CommandType.SELECT, "User Name", null);
+        selectCommand.execute(session, result);
+
+        verify(responder).sendError(eq(session), contains("cannot contain spaces"));
```
[the remark this answers is not in the document]
Comment (TQT279) [reply src/test/java/org/example/VChatTalk/command/implement_command/SelectCommandTest.java]: update in unit test cast contains("Username cannot contain spaces."));

File: src/main/java/org/example/VChatTalk/command/implement_command/ListCommand.java [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+
+    @Override
+    public void execute(WebSocketSession session, CommandResult result) throws IOException {
+        List<String> users = new ArrayList<>();
+
+        for (WebSocketSession s : sessionRegistry.getAllSessions()) {
+            String username = sessionRegistry.getUsername(s.getId());
+            if (username != null && !username.equals("Anonymous")) {
+                users.add(username);
+            }
```
[the remark this answers is not in the document]
Comment (TQT279) [reply src/main/java/org/example/VChatTalk/command/implement_command/ListCommand.java]: using public static final String USER_ANONYMOUS = "Anonymous";

File: src/test/java/org/example/VChatTalk/command/CommandIntegrationTest.java [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+        // 2. Wire up the Commands with the Real Registry
+        List<IChatCommand> commands = Arrays.asList(
+                new LoginCommand(sessionRegistry, responder),
+                new SelectCommand(sessionRegistry, responder),
+                new ListCommand(sessionRegistry, responder),
+                new HelpCommand(responder),
+                new LeaveCommand(sessionRegistry, responder),
+                new ExitCommand(sessionRegistry, responder),
+                new UnknownCommand(responder)
+        );
```
[the remark this answers is not in the document]
Comment (TQT279) [reply src/test/java/org/example/VChatTalk/command/CommandIntegrationTest.java:51]: add comment // Wait until sprint 6 to deploy.

File: src/main/java/org/example/VChatTalk/command/implement_command/SelectCommand.java [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+        // Arg Check
+        String targetUser = result.getArgument();
+        if (targetUser == null || targetUser.isBlank()) {
+            responder.sendError(session, String.format(MessageConstants.ERR_MISSING_ARG_USER, "/select"));
+            return;
+        }
+
+        // Username can not contain blank
+        if (targetUser.trim().contains(" ")) {
+            responder.sendError(session, "Invalid username format. Username cannot contain spaces.");
```
[the remark this answers is not in the document]
Comment (TQT279) [reply src/main/java/org/example/VChatTalk/command/implement_command/SelectCommand.java]: ERR_INVALID_USERNAME is for another err so i adding new ERR_USERNAME_CONTAIN_SPACE to handle this

File: src/main/java/org/example/VChatTalk/command/implement_command/ExitCommand.java [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+    @Override
+    public CommandType getType() {
+        return CommandType.EXIT;
+    }
+
+    @Override
+    public void execute(WebSocketSession session, CommandResult result) throws IOException {
+        String username = sessionRegistry.getUsername(session.getId());
+
+        if (username != null && !username.equals("Anonymous")) {
```
[the remark this answers is not in the document]
Comment (Toffu-BlankieChoccy) [reply src/main/java/org/example/VChatTalk/command/implement_command/ExitCommand.java]: @TQT279 You didn't do anything upon this comment. Why even bother marked as resolved? Should you don't want to resolve, at least give a detailed explanation.

File: src/main/java/org/example/VChatTalk/command/implement_command/ListCommand.java [hunk 10]
```diff
@@ -34,7 +34,7 @@ public void execute(WebSocketSession session, CommandResult result) throws IOExc
 
         for (WebSocketSession s : sessionRegistry.getAllSessions()) {
             String username = sessionRegistry.getUsername(s.getId());
-            if (username != null && !username.equals("Anonymous")) {
+            if (username != null && !username.equals(MessageConstants.USER_ANONYMOUS)) {
```
Comment (Toffu-BlankieChoccy) [inline src/main/java/org/example/VChatTalk/command/implement_command/ListCommand.java:37]: @TQT279  This one is nice but why not do the same thing in ExitCommand?

File: src/main/java/org/example/VChatTalk/command/MessageConstants.java
Comment (Toffu-BlankieChoccy) [reply src/main/java/org/example/VChatTalk/command/MessageConstants.java:1]: > i need permission, so i just send a request to join page jira now need approval to add doc

done

File: src/main/java/org/example/VChatTalk/command/implement_command/ListCommand.java (hunk 10 above)
Comment (TQT279) [reply src/main/java/org/example/VChatTalk/command/implement_command/ListCommand.java:37]: i'm forgot it. i'm just update it in new commit

File: src/main/java/org/example/VChatTalk/command/MessageConstants.java
Comment (TQT279) [reply src/main/java/org/example/VChatTalk/command/MessageConstants.java:1]: i have add MessageConstant Utility to jira pages

**Comment by Toffu-BlankieChoccy** [comment]
All good now. Begin to merge
